### DISCUSSION TITLE
.h files from pragmas and __GNUC__ definition

### DISCUSSION THREAD

**Discussion by MihaMarkic** [BartmanAbyss/vscode-amiga-debug#240] [Q&A] [answered]
I see design time only errors about intellisense not being able to find pragmas/*.h files (referenced in proto/exec.h, dos.h and graphics.h) becasuse `__GNUC__` is not defined at design time I guess. (I don't think pragmas are included in distribution)
What would be a proper way to define it at design time or another way to make VSCode design time happy?

**Comment by BartmanAbyss** [comment]
and this happens with the default project? I can't remember seeing that.

**Reply by MihaMarkic** [reply]
Yep, default project. I guess it might be due to Microsoft C/C++ extension installed.

**Reply by BartmanAbyss** [reply]
I have that extension installed, too. Please tell me what you did exactly (or video) I'm F12-ing around and all is working fine.

**Reply by MihaMarkic** [reply]
First. all except the top include were underlined.
Then I've added `C:\Users\[USER]\.vscode\extensions\bartmanabyss.amiga-debug-1.7.4\bin\win32\opt\m68k-amiga-elf\sys-include` to C/C++ Extension settings' Include path.
Only proto include files are not underlined (all of three, but sometimes one is not for whatever reason).
![image](https://github.com/BartmanAbyss/vscode-amiga-debug/assets/2194935/f71dbc9d-7358-4f8e-a09d-c9d610b4420f)

When I F12 into `proto/exec.h`, I see this
![image](https://github.com/BartmanAbyss/vscode-amiga-debug/assets/2194935/7799452f-e62f-483a-8472-ed8d5eb5068c)

It seems like `__GNUC__` is not defined and intellisense tries to get `pragmas/exec_pragmas.h` which doesn't exist, at least on my computer.

Does this explain it? Let me know if you need more info.

**Reply by BartmanAbyss** [reply]
hmm that's strange. works fine here, you don't need to add any include directories.

**Reply by MihaMarkic** [reply]
Is `__GNUC__` defined for you if you look into `proto/exec.h` file? Do you have exec_pragmas.h on disk? I guess you are more C/C++ oriented and perhaps have those files and global paths somewhere.

**Reply by BartmanAbyss** [reply]
I don't do any C++ development with VSCode. This should all work with a vanilla VSCode installation with vscode-amiga-debug + C/C++ extension.
![image](https://github.com/BartmanAbyss/vscode-amiga-debug/assets/19374186/2d2f45bd-ee93-4ed9-a40f-f48f738b1498)
![image](https://github.com/BartmanAbyss/vscode-amiga-debug/assets/19374186/50e6ed9f-0f80-4602-8803-002f92a233cd)

**Reply by MihaMarkic** [reply]
Mystery thickens. From your photo I see that in your case `_NO_INLINE` seems defined and thus nested block is ignored. I wonder where it is defined.

**Reply by MihaMarkic** [reply]
I also wonder how does C/C++ extension find your .h files in your project. I mean there has to be some place where the path to those is set.

**Reply by MihaMarkic** [reply]
I see you have a  `c_cpp_properties.json` file in your project with `"configurationProvider": "BartmanAbyss.amiga-debug"`. Is it supposed to set the path to your .h files?

**Reply by BartmanAbyss** [reply]
sure, the extension provides a configurationProvider: https://github.com/BartmanAbyss/vscode-amiga-debug/blob/b1f157086acd55c4fca9480f99742c80e8838e37/src/extension.ts#L71-L85

**Reply by MihaMarkic** [reply]
If I had to guess, this is the point where something goes wrong in my case. At least `_NO_INLINE` is not defined for me but probably whole extension method doesn't get called or faults.
I've enabled C/C++ extension logging but don't see any obvious problem. Is there a way to debug call to this method?

**Reply by BartmanAbyss** [reply]
Sure, but you would have to check out the source of the extension and build it yourself. (see README)

**Reply by MihaMarkic** [reply, marked as answer]
Now, this is funny. So I've enabled C/C++ extension logging before and watched the log - nothing obvious there, errors still persisting but nothing related in log. Then I've restarted VSCode to see whether there are errors then, when C/C++ extension is trying to initialize your provider. VSCode did an update (?) and restarted. But this time everything works as expected. I have no idea what did it update (perhaps normal minor version upgrade or some extensions updated).

So I did a few more tests and found this. Intellisense won't work until you restart VSCode. Basically you have to call "Amiga: Init Project", restart VSCode and then intellisense starts working.

**Reply by MihaMarkic** [reply]
Thanks for help :)

**Reply by BartmanAbyss** [reply]
Ah ok, that's interesing. Glad you got it working!
